[CDAP-16236] Using keyvalue widget in runtime arguments modal

itsanudeep opened the pull request:
Now using key value widget in runtime arguments modal.
This would also fix the issues mentioned in https://issues.cask.co/browse/CDAP-16236
Created https://issues.cask.co/browse/CDAP-16292 to write E2E tests that would go into develop later.

itsanudeep committed:
using keyvalue widget in runtime args

ajainarayanan commented:
![Screen Shot 2020-03-02 at 3 50 30 PM](https://user-images.githubusercontent.com/1452845/75728843-e338e900-5c9d-11ea-9397-3847ba13ddfe.png)
@itsanudeep Couple of comments,

1. Using the new widget is ok, but without borders it is hard for me to view/edit. @leacd Can you provide some inputs? I was thinking of the same treatment as the widgets in plugin (border with labels on the top left)
2. Macro keys don't have the disabled cursor when hovering over. Right now I was not sure if it was disabled or if something was amiss on my UI.

ajainarayanan commented on cdap-ui/app/cdap/components/AbstractWidget/AbstractMultiRowWidget/AbstractRow.tsx:
```diff
@@ -41,7 +41,7 @@ export const AbstractRowStyles = (theme): StyleRules => {
 
 export interface IAbstractRowProps<S extends typeof AbstractRowStyles> extends WithStyles<S> {
-  value: string;
+  value: string | any;
```
Can you add a comment as to why this needs `any`?

itsanudeep replied:
done

ajainarayanan commented on cdap-ui/app/cdap/components/AbstractWidget/AbstractMultiRowWidget/index.tsx:
```diff
@@ -151,7 +151,7 @@ export default class AbstractMultiRowWidget<
   };
 
-  public constructValues = () => {
+  public constructValues = (): any => {
```
Same here. We should try avoiding using `any` but if we need to we should provide a comment to justify the necessity.

itsanudeep replied:
done

ajainarayanan commented on cdap-ui/app/cdap/components/PipelineConfigurations/Store/ActionCreator.js:
```diff
@@ -45,36 +45,23 @@ import { CLOUD } from 'services/global-constants';
+    if (pair.key in resolvedMacros) {
       return {
+        notDeletable: true,
+        provided: pair.provided || false,
```
Now that we have removed the checkboxes do we still need the provided?

itsanudeep replied:
fixed

ajainarayanan commented on cdap-ui/app/cdap/components/PipelineConfigurations/Store/ActionCreator.js:
```diff
@@ -45,36 +45,23 @@ import { CLOUD } from 'services/global-constants';
+    return {
+      ...pair,
+      // This is needed because KeyValuePair will render a checkbox only if the provided is a boolean.
+      provided: null,
```
Same here.

itsanudeep replied:
done

ajainarayanan commented on cdap-ui/app/cdap/components/PipelineDetails/PipelineRuntimeArgsDropdownBtn/RuntimeArgsKeyValuePairWrapper/RuntimeArgsPairsMaterial.tsx:
```diff
@@ -0,0 +1,94 @@
+    if (!props.value || props.value.length === 0) {
+      return [];
+    }
+    const kvDelimiter = objectQuery(props, 'widgetProps', 'kv-delimiter') || ':::';
```
Where is this `:::` coming from? I thought we decided to not use a string delimiter as we want to pass in the runtime arguments as is (object).

itsanudeep commented:
- Integrated new runtime arguments widget for use in preview
- Added tests for runtime arguments in preview
- Added tests for runtime arguments on a deployed pipeline

elfenheart commented on cdap-ui/app/cdap/components/AbstractWidget/AbstractMultiRowWidget/index.tsx:
```diff
@@ -68,7 +68,7 @@ export default class AbstractMultiRowWidget<
   };
 
-  private init = (props) => {
+  public init = (props) => {
```
this should follow the pattern on the AbstractRow to have construct and deconstruct methods that needs to be overwritten. It's clearer rather than overwriting init method.

itsanudeep replied:
done

elfenheart commented on cdap-ui/app/cdap/components/PipelineDetails/PipelineRuntimeArgsDropdownBtn/RuntimeArgsKeyValuePairWrapper/RuntimeArgsPairsMaterial.tsx:
```diff
@@ -0,0 +1,96 @@
+    );
+  };
+}
+export default function RuntimeArgsPairs(props) {
```
name component the same as the file name

itsanudeep replied:
done

elfenheart commented on cdap-ui/app/directives/my-pipeline-runtime-args/my-pipeline-runtime-args.html:
```diff
@@ -20,34 +20,14 @@
-    get-resetted-key-value="RuntimeArgsCtrl.getResettedRuntimeArgument"
-  ></key-value-pairs>
+  <key-value-pairs-material
+  widget-props="{
```
nit: indentation

itsanudeep replied:
fixed

elfenheart commented on cdap-ui/app/directives/my-pipeline-runtime-args/my-pipeline-runtime-args.js:
```diff
@@ -46,21 +46,9 @@ class MyPipelineRuntimeArgsCtrl {
-    });
-    this.providedPopoverOpen = !this.providedPopoverOpen;
+  onRuntimeArgumentsChange(changedArgs) {
+    let newArgs = changedArgs.length ? changedArgs : [];
```
nit: const

itsanudeep replied:
fixed

elfenheart commented on cdap-ui/cypress/fixtures/pipeline_with_macros.json:
```diff
@@ -0,0 +1,105 @@
+          "label": "Airport_source",
+          "artifact": {
+            "name": "core-plugins",
+            "version": "2.4.0-SNAPSHOT",
```
use version range for all versions

itsanudeep replied:
fixed

elfenheart commented on cdap-ui/cypress/fixtures/pipeline_with_macros.json:
```diff
@@ -0,0 +1,105 @@
+{
+  "artifact": {
+    "name": "cdap-data-pipeline",
+    "version": "[6.1.0-SNAPSHOT, 6.2.0-SNAPSHOT]",
```
upper bound should be 7.0.0-SNAPSHOT

itsanudeep replied:
fixed

elfenheart reviewed (approved):
minor comment, otherwise 👍 LGTM

itsanudeep commented:
BUT: https://builds.cask.co/browse/CDAP-UDUT505-9
RAT and Checkstyle: https://builds.cask.co/browse/CDAP-DRC9777-9
